routing/probability_bimodal.go

```diff
@@ -0,0 +1,432 @@
+type BimodalConfig struct {
+	// BimodalNodeWeight defines how strongly other channels of the node to
+	// route from should be taken into account. A number between 0 and 1.
+	BimodalNodeWeight float64
+
+	// BimodalScaleMsat describes the scale over which channels still have
+	// some liquidity left on both channel ends.
+	BimodalScaleMsat lnwire.MilliSatoshi
+
+	// DefaultBimodalHalfLife describes the information decay over time for
+	// previous successes or failures.
+	BimodalHalfLife time.Duration
+}
+
+func (p BimodalConfig) validate() error {
+	if p.BimodalHalfLife < 0 {
+		return fmt.Errorf("%v: %w", BimodalEstimatorName,
+			ErrInvalidHalflife)
+	}
+
+	if p.BimodalNodeWeight < 0 || p.BimodalNodeWeight > 1 {
+		return fmt.Errorf("%v: %w", BimodalEstimatorName,
+			ErrInvalidNodeWeight)
+	}
+
+	if p.BimodalScaleMsat == 0 {
+		return fmt.Errorf("%v: %w", BimodalEstimatorName,
+			ErrInvalidScale)
+	}
+
+	return nil
+}
+
+// probabilityBimodal returns node and pair probabilities based on historical
+// payment results based on a liquidity distribution model of the LN.
+type probabilityBimodal struct {
+	// EstimatorConfig contains configuration options for our estimator.
+	EstimatorConfig
+}
+
+// Compile-time check that probabilityBimodal implements Estimator interface.
+var _ Estimator = (*probabilityBimodal)(nil)
+
+// config returns the current configuration of the estimator.
+func (p *probabilityBimodal) config() EstimatorConfig {
+	return p.EstimatorConfig
+}
+
+// String returns the estimator's configuration as a string representation.
+func (p *probabilityBimodal) String() string {
+	return fmt.Sprintf("estimator type: %v, halflife time: %v, liquidity "+
+		"scale: %v, node weight: %v", p.name(), p.BimodalHalfLife,
+		p.BimodalScaleMsat, p.BimodalNodeWeight)
+}
+
+func (p *probabilityBimodal) name() string {
+	return BimodalEstimatorName
+}
+
+// getPairProbability estimates the probability of successfully traversing to
+// toNode based on historical payment outcomes for the from node. Those outcomes
+// are passed in via the results parameter.
+func (p *probabilityBimodal) getPairProbability(now time.Time,
+	results NodeResults, toNode route.Vertex, amt lnwire.MilliSatoshi,
+	capacity btcutil.Amount) float64 {
+
+	// We first compute the probability for the desired hop taking into
+	// account previous knowledge.
+	directProbability := p.getLiquidityProbability(
+		now, results, toNode, amt, lnwire.NewMSatFromSatoshis(capacity),
+	)
+
+	// The final probability is computed by taking into account other
+	// channels of the from node.
+	return p.calculateProbability(directProbability, now, results, toNode)
+}
+
+// getLocalPairProbability computes the probability to reach toNode given a set
+// of previous learnings.
+func (p *probabilityBimodal) getLocalPairProbability(now time.Time,
+	results NodeResults, toNode route.Vertex) float64 {
+
+	// TODO(bitromortac): think more about this case.
+	return 1.0
```

Comment:
IIRC all this needs to tell us is if the channel is online/offline or not. As we want to skip things entirely, since we made the decision to _not_ do any sort of decay for the local channels, as we could end up sort of reducing them all to zero prob, which means no payments attempted at all.